client/src/app/pages/applications/components/application-tags/application-tags.tsx

```diff
@@ -97,48 +110,134 @@ export const ApplicationTags: React.FC<ApplicationTagsProps> = ({
           setIsFetching(false);
         });
     } else {
-      setTagCategories(new Map());
-      setTags(new Map());
+      setTags([]);
+      setTagCategoriesById(new Map());
     }
   }, [application]);
 
-  // TODO(mturley) group by source, with h3 for each source name
+  const sources = new Set<string>();
+  tags.forEach((tag) => sources.add(tag.source || ""));
+
+  const filterCategories: FilterCategory<TagWithSource>[] = [
+    {
+      key: "source",
+      title: t("terms.source"),
+      type: FilterType.multiselect,
+      placeholderText: t("terms.source"),
+      getItemValue: (tag) => tag.source || "Manual",
+      selectOptions: Array.from(sources)
+        .sort(compareSources)
+        .map((source) => source || "Manual")
+        .map((source) => ({ key: source, value: source })),
+      logicOperator: "OR",
+    },
+    {
+      key: "tagCategory",
+      title: t("terms.tagCategory"),
+      type: FilterType.multiselect,
+      placeholderText: t("terms.tagCategory"),
+      getItemValue: (tag) => tag.category?.name || "",
+      selectOptions: Array.from(tagCategoriesById.values())
+        .map((tagCategory) => tagCategory.name)
+        .sort((a, b) => a.localeCompare(b))
+        .map((tagCategoryName) => ({
+          key: tagCategoryName,
+          value: tagCategoryName,
+        })),
+      logicOperator: "OR",
+    },
+  ];
+
+  const {
+    filterValues,
+    setFilterValues,
+    filteredItems: filteredTags,
+  } = useFilterState(tags, filterCategories);
+
+  const tagsBySource = new Map<string, Tag[]>();
+  filteredTags.forEach((tag) => {
+    const tagsInThisSource = tagsBySource.get(tag.source || "");
+    if (tagsInThisSource) {
+      tagsInThisSource.push(tag);
+    } else {
+      tagsBySource.set(tag.source || "", [tag]);
+    }
+  });
 
   return (
     <ConditionalRender when={isFetching} then={<Spinner isSVG size="md" />}>
-      {Array.from(tagCategories.values())
-        .sort((a, b) => (a.rank || 0) - (b.rank || 0))
-        .map((tagCategory) => {
+      <Toolbar
+        clearAllFilters={() => setFilterValues({})}
+        clearFiltersButtonText={t("actions.clearAllFilters")}
+      >
+        <ToolbarContent className={spacing.p_0}>
+          <ToolbarItem>Filter by:</ToolbarItem>
```

Comment:
Ah, @ibolton336 I forgot to put this one through translation. I guess we have to revisit a bunch of those eventually anyway.